### PR TITLE
add create modal optional section

### PR DESCRIPTION
**Pull request by Poltergeist**
---
It is depended on #212 (**merged**)

---

This PR adds the logic for the optional section and also introduces the fields for that section to the modal. 

image

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @jfurrow

File: src/js/components/modals/ServiceFormModal.js [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+];
+
+class ServiceFormModal extends mixin(StoreMixin) {
+  constructor() {
+    super(...arguments);
+
+    let model =
+      ServiceUtil.createFormModelFromSchema(ServiceSchema);
+    this.state = {
+      app: JSON.stringify({id:'', cmd:''}, null, 2),
```
Comment (aldipower) [inline src/js/components/modals/ServiceFormModal.js]: It seems weird to me that the Modal handles the app as a string. ?
Further I think this should come from a store.

File: src/js/components/modals/ServiceFormModal.js [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  handleClearError() {
+    this.setState({
+      errorMessage: null
+    });
+  }
+
+  handleJSONChange(app) {
+    let {service} = this.state;
+    try {
+      service = new Service(JSON.parse(app));
```
Comment (aldipower) [inline src/js/components/modals/ServiceFormModal.js]: The JSON parsing seems to me more the job of a request util?
Feels wrong to have this here.

Comment (aldipower) [inline src/js/components/modals/ServiceFormModal.js]: OK, sorry! It is the JSON mode, now I understand it better.
I would like to see this logic traversed out of this component.
But I am not sure if this is out of scope of this PR. Ignore this comment, if so. :)

File: src/js/components/modals/ServiceFormModal.js [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+          onClick={this.handleSubmit}>
+          Deploy
+        </button>
+      </div>
+    );
+  }
+
+  getModalContents() {
+    let {app, jsonMode, service} = this.state;
+    if (jsonMode) {
```
Comment (aldipower) [inline src/js/components/modals/ServiceFormModal.js]: Still have the feeling the this should be part of another component. But that's another PR.

File: src/js/components/SchemaForm.js [hunk 4]
```diff
@@ -78,7 +85,12 @@ class SchemaForm extends React.Component {
       return;
     }
 
-    setTimeout(() => {
+    // The cleartimeout is there to debounce the validation. And to make
+    // sure it is only run once.
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+    this.timer = setTimeout(() => {
```
Comment (pierlo-upitup) [inline src/js/components/SchemaForm.js]: Can't we use `Util.debounce`?

**Comment by Poltergeist** [comment]
This commit does contain all the commits of #212 please wait with reviewing until #212 is merged.

File: src/js/constants/ServiceSchema.js [hunk 5]
```diff
@@ -81,6 +81,63 @@ let SERVICE_SCHEMA = {
         }
       },
       required: []
+    },
+    'Optional': {
```
Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: Why does this `key` start with a capital letter? Looks very weird.

Comment (pierlo-upitup) [inline src/js/constants/ServiceSchema.js]: it is used as a title of the modal section

File: src/js/constants/ServiceSchema.js [hunk 6]
```diff
@@ -45,7 +45,7 @@ let SERVICE_SCHEMA = {
           type: 'number',
           default: 1,
           getter: function (service) {
-            return (service.getInstancesCount() || 0);
+            return (service.getInstancesCount() || 0) + '';
```
Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: Not sure why this has to be an string, but fair enough, there will be a reason for that.

Comment (pierlo-upitup) [inline src/js/constants/ServiceSchema.js]: Actually if that's the case I'd rather see an ES6 template string here. ````
```
`${service.getInstancesCount() || 0}`
```

Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: Oh, just so that the "type" is set to 'number' above.
So this shouldn't be converted to string here I guess.

File: src/js/utils/ServiceUtil.js [hunk 7]
```diff
@@ -28,6 +28,27 @@ const ServiceUtil = {
         definition.instances = formModel.General.instances;
       }
 
+      if (formModel.Optional != null) {
+        definition.executor = formModel.Optional.executor;
+        definition.fetch = formModel.Optional.uris &&
+          formModel.Optional.uris.split(',')
```
Comment (aldipower) [inline src/js/utils/ServiceUtil.js]: The capital letter in `Optional` looks wrong.

File: src/js/constants/ServiceSchema.js (hunk 6 above)
Comment (pierlo-upitup) [inline src/js/constants/ServiceSchema.js]: Well the `type` there is the `input type="number"` i think. Maybe the string conversion is needed because reading an input's value (`input.value`) would return a string anyway?

Comment (pierlo-upitup) [inline src/js/constants/ServiceSchema.js]: In fact I just saw this on master when opening the dialog: `tracker.js:10 Warning: Failed propType: Invalid prop `startValue` of type `number` supplied to `FieldInput`, expected `string`. Check the render method of `FormControl`.`

Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: Ah ok! So this should be explicitly of type "string"?

Comment (pierlo-upitup) [inline src/js/constants/ServiceSchema.js]: No! It only affects the style of the input, it will show the stepper controls I believe.

File: src/js/utils/ServiceUtil.js [hunk 8]
```diff
@@ -54,6 +75,12 @@ const ServiceUtil = {
...
     appDefinition.cmd = service.getCommand();
 
+    appDefinition.executor = service.getExecutor();
+    appDefinition.fetch = service.getFetch();
+    appDefinition.constraints = service.getConstraints();
+    appDefinition.acceptedResourceRoles = service.getAcceptedResourceRoles();
+    appDefinition.user = service.getUser();
+
     let containerSettings = service.getContainerSettings();
     if (
```
Comment (aldipower) [inline src/js/utils/ServiceUtil.js]: ??

File: src/js/utils/ServiceUtil.js [hunk 9]
```diff
@@ -54,6 +75,12 @@ const ServiceUtil = {
     appDefinition.instances = service.getInstancesCount();
     appDefinition.cmd = service.getCommand();
```
Comment (aldipower) [inline src/js/utils/ServiceUtil.js]: why this extra line at this place?

**Comment by aldipower** [comment]
Could you also remove the extra lines at those places? Thx.
https://github.com/dcos/dcos-ui/pull/267/files#diff-a44f44c46bf6bdbc44737179ffb8970dL2
https://github.com/dcos/dcos-ui/pull/267/files#diff-a44f44c46bf6bdbc44737179ffb8970dR64

File: src/js/constants/ServiceSchema.js (hunk 6 above)
Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: Oh, confused now.
I don't think this "types" here are used to set the input field type.
There's is also a type of `object` and so on here in this scheme.
They are meant to be real javascript types. (Virtually)

File: src/js/utils/ServiceUtil.js (hunk 7 above)
Comment (pierlo-upitup) [inline src/js/utils/ServiceUtil.js]: The `SchemaForm` leverages the key as the title of the section. 
Either way, may I suggest to `let {Optional} = formModel;`? So that we don't have to repeat ourselves so much.

File: src/js/constants/ServiceSchema.js (hunk 6 above)
Comment (pierlo-upitup) [inline src/js/constants/ServiceSchema.js]: I've updated my comments: https://github.com/dcos/dcos-ui/pull/267/files/5b436ccc925cdcf6271d75b71155964504b0d810#r65355596

Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: Oh, so, no convertion to "string". :)

File: src/js/constants/ServiceSchema.js (hunk 5 above)
Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: Sorry, using the key as a value to display is a very bad idea.

Comment (pierlo-upitup) [inline src/js/constants/ServiceSchema.js]: Oh, I totally agree. But that's how the SchemaForm works right now and I think refactoring it to add a `sectionId` or similar is out of scope.

Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: Confirmative.

File: src/js/utils/ServiceUtil.js (hunk 7 above)
Comment (aldipower) [inline src/js/utils/ServiceUtil.js]: I see, confirmative.

File: src/js/constants/ServiceSchema.js (hunk 6 above)
Comment (Poltergeist) [inline src/js/constants/ServiceSchema.js]: the type is as PG mentioned is used for a very simple auto validation but sadly it is not used for the input field. Which means we have to provide a string which is then validated if it is a number. There is definitively some refactoring possible.

File: src/js/utils/ServiceUtil.js (hunk 7 above)
Comment (Poltergeist) [inline src/js/utils/ServiceUtil.js]: Accepted will change, and also do that for the General part above. Should I do that in this PR or do you prefer if we do that in another PR?

File: src/js/utils/ServiceUtil.js (hunk 9 above)
Comment (Poltergeist) [inline src/js/utils/ServiceUtil.js]: Maybe I should add a comment here I thought it would be handy to have a section per section lets say

```JS
// General
definition.id = service.getID();
...
...

// Optional
definition.executor = service.getExecutor();
...
...

File: src/js/constants/ServiceSchema.js [hunk 10]
```diff
@@ -81,6 +81,63 @@ let SERVICE_SCHEMA = {
...
+          description: 'Comma-separated list of resource roles. Marathon ' +
+          'considers only resource offers with roles in this list for ' +
+          'launching tasks of this app.',
+          getter: function (service) {
+            return service.getAcceptedResourceRoles() &&
+              service.getAcceptedResourceRoles().join(', ');
+          }
+        },
+        user: {
+          title: 'user',
```
Comment (pierlo-upitup) [inline src/js/constants/ServiceSchema.js]: `User`

File: src/js/utils/ServiceUtil.js (hunk 8 above)
Comment (Poltergeist) [inline src/js/utils/ServiceUtil.js]: It is a long condition

```JS
    if (
      containerSettings && containerSettings.docker &&
      containerSettings.docker.image
    ) {
```

What we could do here is
```JS

    if (containerSettings && containerSettings.docker &&
      containerSettings.docker.image
    ) {
```
My assumption was that it is more readable as it is.

File: src/js/utils/ServiceUtil.js (hunk 7 above)
Comment (pierlo-upitup) [inline src/js/utils/ServiceUtil.js]: Do it here pls

File: src/js/utils/ServiceUtil.js (hunk 8 above)
Comment (pierlo-upitup) [inline src/js/utils/ServiceUtil.js]: Let's please
```
if (containerSettings && 
  containerSettings.docker &&
  containerSettings.docker.image) {
```

File: src/js/utils/ServiceUtil.js (hunk 9 above)
Comment (aldipower) [inline src/js/utils/ServiceUtil.js]: Ok that makes sense. Comments would be cool.

File: src/js/utils/ServiceUtil.js (hunk 8 above)
Comment (aldipower) [inline src/js/utils/ServiceUtil.js]: I would be fine with everything but the first.

**Comment by Poltergeist** [comment]
What do you think if we pull each section from the schema to its own file and import it separately? This would give us the advantage of not having one huge file.

File: src/js/utils/ServiceUtil.js (hunk 8 above)
Comment (kennyt) [inline src/js/utils/ServiceUtil.js]: we usually do what PG said 👍

File: src/js/utils/ServiceUtil.js (hunk 9 above)
Comment (Poltergeist) [inline src/js/utils/ServiceUtil.js]: Added below this line 😄

File: src/js/constants/service-schema/general.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
+/* eslint-disable no-unused-vars */
+import React from 'react';
+/* eslint-enable no-unused-vars */
+
+let GENERAL = {
```
Comment (aldipower) [inline src/js/constants/service-schema/general.js]: Only "constants" should be written in capital letters.

File: src/js/constants/ServiceSchema.js [hunk 12]
```diff
@@ -2,64 +2,13 @@
 import React from 'react';
 /* eslint-enable no-unused-vars */
 
+import General from './service-schema/general';
```
Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: In general (no I don't mean the var name itself ;) ) the letter casing should follow the file casing.
So I suggest to rename the file `general` to `General`.

File: src/js/constants/ServiceSchema.js [hunk 13]
```diff
@@ -77,7 +26,8 @@ let SERVICE_SCHEMA = {
         }
       },
       required: []
-    }
+    },
+    'Optional': Optional
```
Comment (aldipower) [inline src/js/constants/ServiceSchema.js]: Cool, this decouples it a little bit, I like that.

**Comment by Poltergeist** [comment]
Renamed the files

**Comment by aldipower** [comment]
Nice with ice!

File: src/js/utils/ServiceUtil.js [hunk 14]
```diff
@@ -8,7 +8,7 @@ const getFindPropertiesRecursive = function (service, item) {
 
       return memo;
     }
-    memo[subItem] = item[subItem].getter(service) || item[subItem].default;
+    memo[subItem] = item[subItem].getter && item[subItem].getter(service) || item[subItem].default;
```
Comment (orlandohohmeier) [inline src/js/utils/ServiceUtil.js]: Please use a propper condition instead of boolean operators for this. #nit

**Comment by orlandohohmeier** [comment]
As discussed I'd like to move the schema into a different location (e.g. config) as it's definitely not a constant (any more). Besides that I suggest to make use of the newly [introduced `title` attribute](https://github.com/dcos/dcos-ui/pull/321). But none of this is in the scope of this PR. Let's just keep it in mind while advancing the modal.

**Comment by orlandohohmeier** [comment]
Looks good to me. Great job! Let's 🚢  🇮🇹
